Domains: Add validation for .FR extra info

deBhal opened the pull request:
This PR adds validation for the .FR extra contact information fields. I still need to add validation for the trademark number, but that's fine to go into a separate PR - this one is already Size XL.

This is the client half of a JSON-Schema driven validation model, I'll add more details here tomorrow, but it's client side validation for .FR extra fields, driven by the same JSON Schema as the back end:

![fr-validation](https://user-images.githubusercontent.com/5952255/29671922-246d59e2-892e-11e7-9280-511370831b2f.gif)
(Note: it's even smoother live, the gif makes it look slower and choppier)

akirk commented:
> driven by the same JSON Schema as the back end
For which it is planned to eventually just have it in one place (i.e. the backend).

hewsut commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -62,7 +66,9 @@ class RegistrantExtraInfoFrForm extends React.PureComponent {
 		const translate = this.props.translate;
-		const registrantType = get( this.props.contactDetails, 'extra.registrantType', defaultRegistrantType );
+		const { contactDetails } = this.props;
+		const registrantType = get( contactDetails, 'extra.registrantType', defaultRegistrantType );
```
Eh. 😆  (It's fine. Is that idiomatic? Like never reference beyond `this.props` directly?)

deBhal replied:
No, I just had code here using the `contactDetails` at one point, and forgot to fix it afterwards :/

I have gotten feedback to favour `{ contactDetails } = this.props` over `this.props.contactDetails`, though, and I've grown to appreciate having the props unpacked like that first (it makes it more obvious how to use the component), so I think I will fix this rather than revert it.

hewsut commented on client/components/domains/registrant-extra-info/fr-schema.json:
```diff
@@ -0,0 +1,29 @@
+				}
+			}
+		}
+	},
```
Love the sight of this.  This will be so much win when (heh) we have this for everything.

aidvu commented on client/components/domains/registrant-extra-info/fr-validate-contact-details.js:
```diff
@@ -0,0 +1,38 @@
+ * External dependencies
+ */
+import { get, reduce, set } from 'lodash';
+import validatorFactory from 'is-my-json-valid';
```
TIL, cool! :)

aidvu commented on client/components/domains/registrant-extra-info/fr-validate-contact-details.js:
```diff
@@ -0,0 +1,38 @@
+			// Drop 'data.' prefix
+			const path = String( field ).split( '.' ).slice( 1 );
+
+			return set(
```
Very hard to read, although smart. :laughing:

deBhal replied:
Sorry, thanks :) I had another go, but feel free to object again - I'm familiar with the experience of coming back to code I wrote a month(/week/day) ago and scratching my head ;)

aidvu commented on client/components/domains/registrant-extra-info/fr-validate-contact-details.js:
```diff
@@ -0,0 +1,38 @@
+		validate.errors,
+		( acc, { field, message } ) => {
+			// Drop 'data.' prefix
+			const path = String( field ).split( '.' ).slice( 1 );
```
Maybe better to call it `key` or something? Btw, why the explicit conversion?

deBhal replied:
The `String()`? It's to avoid an error if `field` is `undefined` or something surprising. That only happens if I screw up elsewhere, so it's not _strictly_ necessary, but errors wreck Calypso so hard that I try really hard not to leave openings.

I think `key` would be the right name if we were using it as a single string, like `{ 'extra.sirenSiret': 'validation failed' }` but `path` seems better for this array-of-keys format.  It's what Lodash uses, for example - https://lodash.com/docs/4.17.4#set `_.set( object, path, value )`.

I'm still not totally whether I'm doing this right. I was using it as a key for a while, but it's annoying to consume.

aidvu commented on client/components/domains/registrant-extra-info/fr-form.jsx:
```diff
@@ -116,13 +125,19 @@ class RegistrantExtraInfoFrForm extends React.PureComponent {
 						autoComplete="off"
 						autoCorrect="off"
 						placeholder={ translate( 'ex. XX123456789' ) }
+						isError={ Boolean( validationErrors.registrantVatId ) }
```
There's `formState` for multiple field forms. You can check one https://github.com/Automattic/wp-calypso/blob/8db70cf3a584cd679ed884f25337da3b2140a397/client/my-sites/domains/domain-management/email-forwarding/email-forwarding-add-new.jsx

deBhal replied:
We worked with the `formState` controller on the `contactDetailsForm`, and did some pretty horrible things to be able to use it there, but I don't think it's a particularly good fit here.

Between `Redux` and client-side validation, a lot of it's functionality is unnecessary. Loading, submitting, and handling changes go straight in and out of props, and the validation seems fast enough to just do it(caveat: no performance testing), so there's no need for extra callbacks and tracking `isPendingValidation` et al.

It does mean I'm reinventing the wheel a bit with initializing default values, but unless you strongly disagree, I'd prefer to stick with this for now.

deBhal commented:
> > driven by the same JSON Schema as the back end
> For which it is planned to eventually just have it in one place (i.e. the backend), right? It's just the fields names that need to be transformed from snake_case to camelCase.

Absolutely. We already have exactly what we need in - wpcom.undocumented. `validateGoogleAppsContactInformation()` has literally exactly what we need: `const data = mapKeysRecursively( { contactInformation }, snakeCase );`

deBhal commented:
I've just noticed that using `Input` means I need to mock out local storage, which I'll need to fix before I merge (but not right now it's really late), but that won't materially affect the PR.

kristastevens commented:
Thanks for the request to review copy! Everything looks great.

deBhal commented:
In addition to the store, the `Input` component has some hard-coded events that don't really fit this component, so rather than mock it out I've pulled back from the dependency and added in the validation errors to the form itself.

I've also fixed the schema - it was accepting anything that started with an acceptable pattern.

I still need to add some validation for `trademarkNumber`, but I'll pop that in a separate PR if I can get a thumbs up here.

deBhal commented:
Ok, I've added in the trademark validation, with a little bit of the fancy:
![fr-validation-multiple-messages](https://user-images.githubusercontent.com/5952255/29871916-d61070c0-8dd0-11e7-9918-47ceaf749728.gif)

hewsut commented:
Looking good. Two notes so far from testing:

1. The SIREN/SIRET and Trademark fields accept (mathematical) "e" input which makes them feel buggy from a user POV if you pound a bunch of keys or start retyping the example text (ie, "ex. 1393943"). I think that we need to find a way to disallow the "e" to proceed.

2. The SIREN/SIRET example has spaces but the field doesn't accept spaces. It'd be nice if they were congruous one direction or the other. I think the canonical direction is with the spacing? Not a total show stopper but unexpected.

hewsut commented:
Mulling. Should we cap the SIREN/SIRET field at 14 and Trademark at 9? Why accept over?

deBhal commented:
> Should we cap the SIREN/SIRET field at 14 and Trademark at 9? Why accept over?

I think it's slightly easier to fix if we allow the extra characters. If you butter-fingers an extra number in somewhere you can go and delete that character, or if you confused one field for another you can paste it into the right field.

I wonder if that's an argument for allowing alpha-numeric characters in the SIREN/SIRET field - they're all very closely related for French businesses (SIRET and French VAT ID both include SIREN).

deBhal commented:
I've added sanitization for the two numeric fields and switched them to `type="text"` to avoid the unpleasant behaviour around `e`, `.` and `-`.  The good news is that mobile devices will still pick up on the `pattern="[0-9]*"` and show the right keyboard.

I wrote come code to format the field, but it resulting in an annoying bug with the cursor jumping to the end on every edit. I've backed that out for now, and it can go into a different PR, I think.
